client/post-editor/media-modal/image-editor/image-editor-canvas.jsx

```diff
@@ -67,78 +82,136 @@ const MediaModalImageEditorCanvas = React.createClass( {
+		// setTransform() could be replaced with translate(), but it leads to
+		// a false positive warning from eslint rule wpcalypso/i18n-no-variables
 		context.setTransform( 1, 0, 0, 1, canvas.width / 2, canvas.height / 2 );
 
-		context.scale( this.props.scaleX, this.props.scaleY );
-		context.rotate( this.props.degrees * Math.PI / 180 );
-		context.drawImage( this.image, -this.image.width / 2, -this.image.height / 2 );
+		context.scale( transform.scaleX, transform.scaleY );
+		context.rotate( transform.degrees * Math.PI / 180 );
+
+		context.drawImage( this.image, -imageWidth / 2, -imageHeight / 2 );
+
 		context.restore();
 	},
 
+	updateCanvasPosition() {
+		const canvas = ReactDom.findDOMNode( this.refs.canvas );
+		const canvasX = - 50 * this.props.crop.widthRatio - 100 * this.props.crop.leftRatio;
+		const canvasY = - 50 * this.props.crop.heightRatio - 100 * this.props.crop.topRatio;
+
+		this.props.setImageEditorCropBounds(
+			canvas.offsetTop - canvas.offsetHeight * -canvasY / 100,
+			canvas.offsetLeft - canvas.offsetWidth * -canvasX / 100,
+			canvas.offsetTop + canvas.offsetHeight * ( 1 + canvasY / 100 ),
+			canvas.offsetLeft + canvas.offsetWidth * ( 1 + canvasX / 100 ) );
+	},
+
+	preventDrag( event ) {
+		event.preventDefault();
+		return false;
+	},
+
+	renderCrop() {
+		if ( ! this.props.src ) {
+			return;
+		}
+
+		return ( <Crop /> );
+	},
+
 	render() {
-		const rotatedMod = this.props.degrees % 180,
-			width = rotatedMod === 0 ? this.state.canvasWidth : this.state.canvasHeight,
-			height = rotatedMod === 0 ? this.state.canvasHeight : this.state.canvasWidth;
+		const canvasX = - 50 * this.props.crop.widthRatio - 100 * this.props.crop.leftRatio;
+		const canvasY = - 50 * this.props.crop.heightRatio - 100 * this.props.crop.topRatio;
+
+		const canvasStyle = {
+			top: 50 + '%',
+			left: 50 + '%',
+			transform: 'translate(' + canvasX + '%, ' + canvasY + '%)',
+			maxWidth: ( 85 / this.props.crop.widthRatio ) + '%',
+			maxHeight: ( 85 / this.props.crop.heightRatio ) + '%'
+		};
 
 		return (
 			<div className="editor-media-modal-image-editor__canvas-container">
+				{ this.renderCrop() }
 				<canvas
 					ref="canvas"
-					className="editor-media-modal-image-editor__canvas"
-					width={ width }
-					height={ height } />
+					style={ canvasStyle }
+					onMouseDown={ this.preventDrag }
+					className="editor-media-modal-image-editor__canvas" />
 			</div>
 		);
 	}
 } );
 
 export default connect(
 	( state ) => {
-		const { degrees, scaleX, scaleY } = getImageEditorTransform( state ),
-			{ src, mimeType } = getImageEditorFileInfo( state );
+		const transform = getImageEditorTransform( state );
+		const { src, mimeType } = getImageEditorFileInfo( state );
+		const crop = getImageEditorCrop( state );
 
 		return {
 			src,
 			mimeType,
-			degrees,
-			scaleX,
-			scaleY
+			transform,
+			crop
 		};
 	},
-	null,
+	{ setImageEditorCropBounds },
```

Comment:
>Why did you decide to pass the function as a prop, instead of using the imported variable directly in the lower-level component?

Not obvious at a glance, but the prop that's injected into the component here is wrapped with a `dispatch` call to the context `store`:

> If an object is passed, each function inside it will be assumed to be a Redux action creator. An object with the same function names, but with every action creator wrapped into a `dispatch` call so they may be invoked directly, will be merged into the component’s props.

https://github.com/reactjs/react-redux/blob/master/docs/api.md#connectmapstatetoprops-mapdispatchtoprops-mergeprops-options